spec.emu

```diff
@@ -23,83 +23,142 @@ contributors: Krzysztof Kotowicz, Mike Samuel
+      1. Let _exprSym_ be the grammar symbol |GeneratorExpression|.
+      1. Let _bodySym_ be the grammar symbol |GeneratorBody|.
+      1. Let _parameterSym_ be the grammar symbol |FormalParameters[+Yield, ~Await]|.
       1. Let _fallbackProto_ be *"%GeneratorFunction.prototype%"*.
     1. Else if _kind_ is ~async~, then
-      1. Let _goal_ be the grammar symbol |AsyncFunctionBody|.
-      1. Let _parameterGoal_ be the grammar symbol |FormalParameters[~Yield, +Await]|.
+      1. Let _prefix_ be *"async function"*.
+      1. Let _exprSym_ be the grammar symbol |AsyncFunctionExpression|.
+      1. Let _bodySym_ be the grammar symbol |AsyncFunctionBody|.
+      1. Let _parameterSym_ be the grammar symbol |FormalParameters[~Yield, +Await]|.
       1. Let _fallbackProto_ be *"%AsyncFunction.prototype%"*.
     1. Else,
-      1. Assert: _kind_ is ~asyncGenerator~.
-      1. Let _goal_ be the grammar symbol |AsyncGeneratorBody|.
-      1. Let _parameterGoal_ be the grammar symbol |FormalParameters[+Yield, +Await]|.
+      1. Assert: _kind_ is ~async-generator~.
+      1. Let _prefix_ be *"async function\*"*.
+      1. Let _exprSym_ be the grammar symbol |AsyncGeneratorExpression|.
+      1. Let _bodySym_ be the grammar symbol |AsyncGeneratorBody|.
+      1. Let _parameterSym_ be the grammar symbol |FormalParameters[+Yield, +Await]|.
       1. Let _fallbackProto_ be *"%AsyncGeneratorFunction.prototype%"*.
-    1. Let _argCount_ be the number of elements in _args_.
-    1. Let _P_ be the empty String.
+    1. Let _argCount_ be the number of elements in _parameterArgs_.
     1. <ins>Let _assembledFromCodeLike_ be *true*.</ins>
-    1. If _argCount_ = 0, let _bodyArg_ be the empty String.
-    1. Else if _argCount_ = 1, <del>let _bodyArg_ be _args_[0]</del>.
-      1. <ins>Let _bodyArg_ be _args_[0].</ins>
-      1. <ins>If IsCodeLike(_bodyArg_) is *true*, set _bodyArg_ to _bodyArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
-    1. Else,
-      1. Assert: _argCount_ &gt; 1.
-      1. Let _firstArg_ be _args_[0].
-      1. <ins>If IsCodeLike(_firstArg_) is *true*, set _firstArg_ to _firstArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
-      1. Set _P_ to ? ToString(_firstArg_).
+    1. <ins>If IsCodeLike(_bodyArg_) is *true*, set _bodyArg_ to _bodyArg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
+    1. Let _bodyString_ be ? ToString(_bodyArg_).
+    1. Let _parameterStrings_ be a new empty List.
+    1. For each element _arg_ of _parameterArgs_, do
+      1. <ins>If IsCodeLike(_arg_) is *true*, set _arg_ to _arg_.[[HostDefinedCodeLike]]. Otherwise, set _assembledFromCodeLike_ to *false*.</ins>
+      1. Append ? ToString(_arg_) to _parameterStrings_.
+    1. Let _currentRealm_ be the current Realm Record.
+    // TODO: HostValidateDynamicCode needs to be changed to handle the new model where body and parameter strings are separated.
```

Comment:
Need to update this algorithm such that HostValidateDynamicCode returning the new source text can be merged with the current flow where things are separated.

It seems that a large chunk of the algorithm is converting from the multiple args to a single source string. It's possible that needs to happen *before* calling. HostValidateDynamicCode, that way it can take the whole string itself and also return a solid string too?